storage/innobase/row/row0purge.cc

```diff
@@ -271,6 +271,433 @@ row_purge_remove_clust_if_poss(
+				}
+
+				entry = row_build_index_entry(
+					row, ext, index, heap);
+				if (entry && !dtuple_coll_cmp(ientry, entry)) {
+					goto unsafe_to_purge;
+				}
+			} else {
+				/* Build index entry out of row */
+				entry = row_build_index_entry(row, ext, index, heap);
+				/* entry could only be NULL if
+				the clustered index record is an uncommitted
+				inserted record whose BLOBs have not been
+				written yet. The secondary index record
+				can be safely removed, because it cannot
+				possibly refer to this incomplete
+				clustered index record. (Insert would
+				always first be completed for the
+				clustered index record, then proceed to
+				secondary indexes.) */
+
+				if (entry && row_purge_vc_matches_cluster(
+					    rec, entry,
+					    clust_index, clust_offsets,
+					    index, ientry, roll_ptr,
+					    trx_id, mtr, &vrow)) {
+					goto unsafe_to_purge;
+				}
+			}
+			clust_offsets = rec_get_offsets(rec, clust_index, NULL,
+							clust_index
+							->n_core_fields,
+							ULINT_UNDEFINED, &heap);
+		} else {
+
+			entry = row_build_index_entry(
+				row, ext, index, heap);
+
+			/* If entry == NULL, the record contains unset BLOB
+			pointers.  This must be a freshly inserted record.  If
+			this is called from
+			row_purge_remove_sec_if_poss_low(), the thread will
+			hold latches on the clustered index and the secondary
+			index.  Because the insert works in three steps:
+
+				(1) insert the record to clustered index
+				(2) store the BLOBs and update BLOB pointers
+				(3) insert records to secondary indexes
+
+			the purge thread can safely ignore freshly inserted
+			records and delete the secondary index record.  The
+			thread that inserted the new record will be inserting
+			the secondary index records. */
+
+			/* NOTE that we cannot do the comparison as binary
+			fields because the row is maybe being modified so that
+			the clustered index record has already been updated to
+			a different binary value in a char field, but the
+			collation identifies the old and new value anyway! */
+			if (entry && !dtuple_coll_cmp(ientry, entry)) {
+unsafe_to_purge:
+				mem_heap_free(heap);
+
+				if (v_heap) {
+					mem_heap_free(v_heap);
+				}
+				return true;
+			}
+		}
+	} else if (dict_index_has_virtual(index)) {
+		/* The current cluster index record could be
+		deleted, but the previous version of it might not. We will
+		need to get the virtual column data from undo record
+		associated with current cluster index */
+
+		cur_vrow = row_vers_build_cur_vrow(
+			rec, clust_index, &clust_offsets,
+			index, trx_id, roll_ptr, heap, v_heap, mtr);
+	}
+
+	version = rec;
+
+	for (;;) {
+		heap2 = heap;
+		heap = mem_heap_create(1024);
+		vrow = NULL;
+
+		trx_undo_prev_version_build(version,
+					    clust_index, clust_offsets,
+					    heap, &prev_version, mtr,
+					    TRX_UNDO_CHECK_PURGE_PAGES,
+					    nullptr,
+					    dict_index_has_virtual(index)
+					    ? &vrow : nullptr);
```

Comment:
I implemented this in b97818a81ac32fe348867e797d5fe8c5f3f466e0 and it initially seemed to be fine (up to a 12-second workload). I was still able reproduce the infinite loop after a 60-second or longer benchmark run. After kill, recovery, and another slow shutdown, that data directory recovered fine. @mariadb-DebarunBanerjee do you have any ideas?

Reply:
If I kill the server, make a copy of the data directory, start and shut down with `innodb_fast_shutdown=0`, the undo log will be purged just fine. I repeated this a couple of times. I was unable to reproduce the hang with `rr` so far. I also tried a debug instrumented executable, and it did not produce any assertion failures when it occasionally hit the infinite loop.

Reply:
In the follow-up commit 945f9442e0f1f089db4182f5e33abb89ba5f9850 I made `row_vers_impl_x_locked_low()` release the clustered index page latch during the traversal of the history. This I now think could be unsafe, because a new updated version of the clustered index record could actually start to match the secondary index record. The scenario is that the value of the indexed column is being updated back and forth between a few values. This may also be the explanation why b97818a81ac32fe348867e797d5fe8c5f3f466e0 causes the deadlocks. The infinite loop would then start already while the SQL workload is running.

Similarly, in `row_check_index()` (`CHECK TABLE`) maybe we’d better hold the clustered index page latch for the entire duration of checking the versions.

Reply:
45a957b0410d6c86fa7b004a3bade088ef8e9e35 reverts these controversial changes.

Reply:
The actual bottleneck seems to be a too high value of `innodb_purge_batch_size`. For the single-table, single-row update workload in [MDEV-34515](https://jira.mariadb.org/browse/MDEV-34515), on my system the maximum good value seems to be `innodb_purge_batch_size=128`. If I increase it from there, the throughput will drop to less than 10%.